pallets/nodle-staking/src/lib.rs

```diff
@@ -0,0 +1,2311 @@
+mod benchmarking;
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+mod set;
+pub mod weights;
+
+use frame_support::pallet;
+mod migrations;
+pub mod slashing;
+
+#[cfg(feature = "std")]
+use frame_support::traits::GenesisBuild;
+
+pub use pallet::*;
+
+#[pallet]
+pub mod pallet {
+    use super::*;
+    use crate::set::OrderedSet;
+    use frame_support::{
+        pallet_prelude::*,
+        traits::{
+            Currency, ExistenceRequirement, Get, Imbalance, LockIdentifier, LockableCurrency,
+            OnUnbalanced, WithdrawReasons,
+        },
+    };
+    use frame_system::pallet_prelude::*;
+    use frame_system::{self as system};
+    use pallet_session::historical;
+    use parity_scale_codec::{Decode, Encode, HasCompact};
+    use sp_runtime::{
+        traits::{AccountIdConversion, AtLeast32BitUnsigned, Convert, Saturating, Zero},
+        DispatchResult, ModuleId, Perbill, RuntimeDebug,
+    };
+    use sp_staking::{
+        offence::{Offence, OffenceDetails, OffenceError, OnOffenceHandler, ReportOffence},
+        SessionIndex,
+    };
+    use sp_std::{cmp::Ordering, convert::From, prelude::*};
+
+    pub use weights::WeightInfo;
+
+    pub(crate) const STAKING_ID: LockIdentifier = *b"staking ";
+
+    pub const MAX_UNLOCKING_CHUNKS: usize = 32;
+
+    #[derive(Default, Clone, Encode, Decode, RuntimeDebug)]
+    pub struct Bond<AccountId, Balance> {
+        pub owner: AccountId,
+        pub amount: Balance,
+    }
+
+    impl<A, B: Default> Bond<A, B> {
+        fn from_owner(owner: A) -> Self {
+            Bond {
+                owner,
+                amount: B::default(),
+            }
+        }
+    }
+
+    impl<AccountId: Ord, Balance> Eq for Bond<AccountId, Balance> {}
+
+    impl<AccountId: Ord, Balance> PartialEq for Bond<AccountId, Balance> {
+        fn eq(&self, other: &Self) -> bool {
+            self.owner == other.owner
+        }
+    }
+
+    impl<AccountId: Ord, Balance> Ord for Bond<AccountId, Balance> {
+        fn cmp(&self, other: &Self) -> Ordering {
+            self.owner.cmp(&other.owner)
+        }
+    }
+
+    impl<AccountId: Ord, Balance> PartialOrd for Bond<AccountId, Balance> {
+        fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+            Some(self.cmp(other))
+        }
+    }
+
+    /// Just a Balance/BlockNumber tuple to encode when a chunk of funds will be unlocked.
+    #[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, Copy)]
+    pub struct UnlockChunk<Balance> {
+        /// Amount of funds to be unlocked.
+        value: Balance,
+        /// Session number at which point it'll be unlocked.
+        session_idx: SessionIndex,
+    }
+
+    #[derive(Copy, Clone, PartialEq, Eq, Encode, Decode, RuntimeDebug)]
+    /// The activity status of the validator
+    pub enum ValidatorStatus {
+        /// Committed to be online and producing valid blocks
+        Active,
+        /// Temporarily inactive
+        Idle,
```

Comment:
Do we still support validators to chill or become idle?

Reply:
Under on_offence or slashing context, Validator is removed from the pool & blocking them from Validator selection process, In this state Validator is moved to "Idle", but bond is still locked.

To get out of this state & move to "Active" state back, validator needs to call validator_bond_more() with Some bond value. SO validator can join in validator selection process from next session.